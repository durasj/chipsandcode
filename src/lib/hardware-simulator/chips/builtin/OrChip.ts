import type { ChipPin } from '../Chip';
import type Chip from '../Chip';
import IllegalStateError from '../IllegalStateError';

/**
 * OR gate implemented natively
 */
class OrChip implements Chip {
  public readonly name = 'Or';

  private a = false;
  private b = false;
  private out = false;

  setInput(name: 'a' | 'b', value: boolean) {
    if (name !== 'a' && name !== 'b')
      throw new IllegalStateError(`Input pin '${name}' does not exist on 'Or'. Input pins: a, b.`);

    this[name] = value;
  }

  getOutput(name: string) {
    if (name !== 'out')
      throw new IllegalStateError(`Output pin '${name}' does not exist on 'Or'. Output pins: out.`);
    return this.out;
  }

  getPins(): Map<string, ChipPin> {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['a', { type: 'input', state: this.a, connections: [] }],
      ['b', { type: 'input', state: this.b, connections: [] }],
      ['out', { type: 'output', state: this.out, connections: [] }],
    ]);
  }

  public run() {
    this.out = this.a || this.b;
  }
}

export default OrChip;
