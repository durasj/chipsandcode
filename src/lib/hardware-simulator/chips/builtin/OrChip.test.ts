import { describe, expect, it } from 'vitest';

import OrChip from './OrChip';

describe('Or Chip', () => {
  it('Provides interface info', async () => {
    const chip = new OrChip();

    expect(chip.name).toBe('Or');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new OrChip();

    const values = [
      [false, false, false],
      [true, false, true],
      [false, true, true],
      [true, true, true],
    ];

    values.forEach(([a, b, out]) => {
      chip.setInput('a', a);
      chip.setInput('b', b);
      chip.run();
      expect(chip.getOutput('out')).toBe(out);
    });
  });

  it('Provides helpful error if setting input on an unknown pin', () => {
    const chip = new OrChip();

    // @ts-ignore We are testing incorrect input on purpose
    expect(() => chip.setInput('c', true)).toThrowError(
      `Input pin 'c' does not exist on 'Or'. Input pins: a, b.`,
    );

    // @ts-ignore We are testing incorrect input on purpose
    expect(() => chip.setInput('out', true)).toThrowError(
      `Input pin 'out' does not exist on 'Or'. Input pins: a, b.`,
    );
  });

  it('Provides helpful error if getting output from an unknown pin', () => {
    const chip = new OrChip();

    expect(() => chip.getOutput('b')).toThrowError(
      `Output pin 'b' does not exist on 'Or'. Output pins: out.`,
    );

    expect(() => chip.getOutput('a')).toThrowError(
      `Output pin 'a' does not exist on 'Or'. Output pins: out.`,
    );
  });
});
