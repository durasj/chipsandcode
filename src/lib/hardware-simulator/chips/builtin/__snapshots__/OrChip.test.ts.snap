// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Or Chip > Provides interface info 1`] = `
Map {
  "a" => {
    "connections": [],
    "state": false,
    "type": "input",
  },
  "b" => {
    "connections": [],
    "state": false,
    "type": "input",
  },
  "out" => {
    "connections": [],
    "state": false,
    "type": "output",
  },
}
`;
